### DISCUSSION TITLE
Bot is slow compared to other libraries on different languages

### DISCUSSION THREAD

**Discussion by DTraitor** [reo7sp/tgbot-cpp#285] [Q&A] [answered]
Hi, guys!
I spent some time looking for C++ libraries for telegram bots and found this. After doing some tests I noticed that it's generally slower than the bot written in Python with a library I used before (python-telegram-bot). For example sending messages using tgbot-cpp takes between 0.200 and 0.280 seconds while in python it takes between 0.077 and 0.190 seconds.
Am I missing something or it's how tgbot-cpp works? (not trying to insult anyone, just wondering)

P.S. I compared a **release** build of the bot
P.S.S. You can notice the difference in performance as a user when waiting for the bot message

**Comment by DTraitor** [comment]
I did some testing and after replacing boost::property_tree with rapidjson the bot actually became slower. Another possible reason for this is the use of pointers almost in every object

**Comment by reo7sp** [comment]
>0.200 seconds, 0.280 seconds

It looks like you’re measuring time with network. Network time is volatile, so it’s not so proper comparison. 

But if you want to continue research, there’s a way to compare - make a lot of measurements of time and calculate percentiles (for example 95th). Percentiles help overcome volatile nature of network time.

Refactor from boost to rapidjson won’t give much nor performance improvement, nor degradation, because json parsing time is nanosecond scale while network time is microsecond scale

**Comment by reo7sp** [comment, marked as answer]
The measured 77-280 ms includes DNS/TCP/TLS, network routing, Telegram processing and the response transfer, so it does not isolate language or JSON performance. Current `CurlHttpClient` keeps one libcurl handle per thread, which allows connection reuse on repeated requests from the same thread.

For a meaningful comparison:

1. Run both bots on the same machine and network.
2. Warm up both clients before measuring.
3. Send the same Bot API method with the same payload sequentially.
4. Record hundreds of requests and compare median, p95 and p99, not one or two samples.
5. Log libcurl timing or measure the HTTP call separately from update handling.

If the difference remains, include the benchmark code, sample count, percentiles and whether both clients reuse a connection. The original timings alone do not demonstrate a tgbot-cpp defect.
